### PR TITLE
Beginning implementation of distributed rfep

### PR DESCRIPTION
**Pull request by pgrinaway**
Initial start of distributed tasklets for nonequilibrium switching. This is not quite ready to merge yet--it's only intended to determine exactly what is needed to get things working in Celery. Thus far, it does work, but we'll want to work through the API a little bit more.

### CODE REVIEW

**Comment by jchodera** [comment]
Awesome!

File: examples/relative_hydration.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    initial_reverse_positions = initial_state.getPositions(asNumpy=True)
+    equil_positions = simulate_hybrid(hybrid_system,functions, 1.0, initial_reverse_positions, nsteps=1000)
+
+    #now, reverse protocol
+    print("Beginning reverse protocols...")
+    with progressbar.ProgressBar(max_value=n_iterations) as bar:
+        for i in range(n_iterations):
+            equil_positions = simulate_hybrid(hybrid_system,functions, 1.0, equil_positions)
+            reverse_context.setPositions(equil_positions)
+            reverse_integrator.step(ncmc_nsteps)
```
Comment (jchodera) [inline examples/relative_hydration.py]: I think have to reset the `reverse_integrator` each iteration with a call to [`reverse_integrator.reset()`](https://github.com/choderalab/perses/blob/master/perses/annihilation/ncmc_switching.py#L954-L967).

File: examples/relative_hydration.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+
+kB = unit.BOLTZMANN_CONSTANT_kB * unit.AVOGADRO_CONSTANT_NA
+temperature = 300.0 * unit.kelvin
+kT = kB * temperature
+beta = 1.0/kT
+
+def simulate_hybrid(hybrid_system,functions, lambda_value, positions, nsteps=500, timestep=1.0*unit.femtoseconds, temperature=temperature, collision_rate=5.0/unit.picoseconds):
+    platform = openmm.Platform.getPlatformByName("CUDA")
+    integrator = openmm.LangevinIntegrator(temperature, collision_rate, timestep)
```
Comment (jchodera) [inline examples/relative_hydration.py]: I think you want to use the `openmmtools.integrators.LangevinIntegrator` or `GHMCIntegrator` instead here.

File: examples/relative_hydration.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    initial_forward_positions = initial_state.getPositions(asNumpy=True)
+    equil_positions = simulate_hybrid(hybrid_system,functions, 0.0, initial_forward_positions)
+
+    print("Beginning forward protocols")
+    #first, do forward protocol (lambda=0 -> 1)
+    with progressbar.ProgressBar(max_value=n_iterations) as bar:
+        for i in range(n_iterations):
+            equil_positions = simulate_hybrid(hybrid_system, functions, 0.0, equil_positions)
+            forward_context.setPositions(equil_positions)
+            forward_integrator.step(ncmc_nsteps)
```
Comment (jchodera) [inline examples/relative_hydration.py]: Same as below: You have to call [`forward_integrator.reset()`](https://github.com/choderalab/perses/blob/master/perses/annihilation/ncmc_switching.py#L954-L967) each iteration.

**Comment by jchodera** [comment]
@pgrinaway: Can we merge in this PR and tackle remaining issues in separate PRs targeted to the issues that are possible to review rapidly and individually?

**Comment by pgrinaway** [comment]
I'm cleaning up the repo now. I'm not sure whether this should be merged separately, since a number of commits are in #352 as well. Reconciling that this morning.

**Comment by pgrinaway** [comment]
Ok, looks like this is only one commit ahead of the other, so we can merge this one after review.

**Comment by jchodera** [comment]
See @Lnaden and the #yank channel on slack for info on how to work around the segfault by moving the openmm dev package tests to a separate optional travis branch and using release openmm for standard py27/35/36 tests.

I'll look through the PR now.

**Comment by Lnaden** [comment]
You can also specifically look at choderalab/openmmtools#275 to see how I fixed it in the .travis.yml file

File: perses/distributed/feptasks.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+@app.task(bind=True, base=NonequilibriumSwitchTask, serializer="pickle")
+def run_equilibrium(self, starting_positions, nsteps, lambda_state, functions, thermodynamic_state, integrator):
+    """
+    Run nsteps of equilibrium sampling at the specified thermodynamic state and return the positions.
+
+    Returns
+    -------
+    positions : [n, 3] np.ndarray quantity
+    """
+    integrator = restore_properties(integrator)
```
Comment (pgrinaway) [inline perses/distributed/feptasks.py]: No, it just adds properties back. It's a temporary hack to get around the problem of not returning the Python type.

File: perses/distributed/feptasks.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    Perform a switching protocol and return the nonequilibrium switching weight
+
+    Returns
+    -------
+    weight : float64
+        The nonequilibrium switching weight
+    """
+    integrator = restore_properties(integrator)
+    switching_ctx, integrator_neq = self._cache.get_context(thermodynamic_state, integrator)
```
Comment (pgrinaway) [inline perses/distributed/feptasks.py]: I did change this, but it's in another branch, which is why I wanted to just merge that one.

File: examples/relative_hydration.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+    platform = openmm.Platform.getPlatformByName("CUDA")
+
+    hybrid_system = factory.hybrid_system
+    hybrid_topology = factory.hybrid_topology
+    initial_hybrid_positions = factory.hybrid_positions
+
+
+    #alchemical functions
+    functions = {
```
Comment (pgrinaway) [inline examples/relative_hydration.py]: I did. We should merge a different branch. It could be pretty annoying to reconcile all of those commits into this

**Comment by pgrinaway** [comment]
Commits here are contained in #380. Closing this PR.

**Pull request closed without merging**